### PR TITLE
🎉 New Source: todoist

### PR DESCRIPTION
**Pull request by netoferraz**
## What

- Added a source for the [todoist](https://developer.todoist.com/rest/v2/) REST API.

## How
Using Airbyte Python CDK

## Recommended reading order
1. `airbyte-integrations/connectors/source-todoist/source_todoist/spec.yaml`
2. `airbyte-integrations/connectors/source-todoist/source_todoist/source.py`

## 🚨 User Impact 🚨
No breaking change

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

image

</details>

<details><summary><strong>Integration</strong></summary>

image
</details>

<details><summary><strong>Acceptance</strong></summary>

image

</details>

### CODE REVIEW

**Comment by sajarin** [comment]
Thanks for the pr @netoferraz someone will be assigned to take a look soon!

**Comment by delenamalan** [comment]
Happy to review soon :)

File: airbyte-integrations/connectors/source-todoist/unit_tests/test_streams.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+import pytest
+from source_todoist.source import TodoistStream
+
+
+@pytest.fixture
+def patch_base_class(mocker):
+    # Mock abstract methods to enable instantiating abstract class
+    mocker.patch.object(TodoistStream, "path", "v0/example_endpoint")
+    mocker.patch.object(TodoistStream, "primary_key", "test_primary_key")
```
Comment (delenamalan) [inline airbyte-integrations/connectors/source-todoist/unit_tests/test_streams.py:16]: You can remove these if you're not using them.

**Review by deanmalan — CHANGES_REQUESTED** [review]
Nice work, thank you @netoferraz! I added a few minor comments.

File: airbyte-integrations/connectors/source-todoist/source_todoist/spec.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+documentationUrl: https://docs.airbyte.io/integrations/sources/todoist
+connectionSpecification:
+  $schema: http://json-schema.org/draft-07/schema#
+  title: todoist REST API
```
[the remark this answers is not in the document]
Comment (netoferraz) [reply airbyte-integrations/connectors/source-todoist/source_todoist/spec.yaml]: done.

File: airbyte-integrations/connectors/source-todoist/unit_tests/test_streams.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    mocker.patch.object(TodoistStream, "__abstractmethods__", set())
+
+
+def test_request_params(patch_base_class):
+    stream = TodoistStream()
+    # TODO: replace this with your input parameters
```
[the remark this answers is not in the document]
Comment (netoferraz) [reply airbyte-integrations/connectors/source-todoist/unit_tests/test_streams.py]: done.

File: airbyte-integrations/connectors/source-todoist/schemas/projects.json [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    "name": {
+      "type": "string"
+    },
+    "order": {
+      "type": "integer"
+    },
+    "parent_id": {
+      "type": "null"
+    },
```
[the remark this answers is not in the document]
Comment (netoferraz) [reply airbyte-integrations/connectors/source-todoist/schemas/projects.json]: done.

File: docs/integrations/sources/todoist.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# Todoist
+
+## Overview
+
+The Todoist source supports only `Full Refresh` syncs.
+
+### Output schema
+
+Two output streams are available from this source. A list of these streams can be found below in the [Streams](outreach.md#streams) section.
```
[the remark this answers is not in the document]
Comment (netoferraz) [reply docs/integrations/sources/todoist.md]: done.

File: docs/integrations/sources/todoist.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+| Feature           | Supported? |
+|:------------------|:-----------|
+| Full Refresh Sync | Yes        |
+| Incremental Sync  | No         |
+
+## Getting started
+
+### Requirements
+
+* Token API
```
[the remark this answers is not in the document]
Comment (netoferraz) [reply docs/integrations/sources/todoist.md]: done.

File: docs/integrations/sources/todoist.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+### Setup guide
```
[the remark this answers is not in the document]
Comment (netoferraz) [reply docs/integrations/sources/todoist.md]: done.

File: airbyte-integrations/connectors/source-todoist/integration_tests/abnormal_state.json [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{
+  "todo-stream-name": {
```
[the remark this answers is not in the document]
Comment (netoferraz) [reply airbyte-integrations/connectors/source-todoist/integration_tests/abnormal_state.json]: done.

File: airbyte-integrations/connectors/source-todoist/schemas/projects.json [hunk 9]
```diff
@@ -0,0 +1,42 @@
+{
```
[the remark this answers is not in the document]
Comment (netoferraz) [reply airbyte-integrations/connectors/source-todoist/schemas/projects.json]: done.

File: airbyte-integrations/connectors/source-todoist/unit_tests/test_streams.py (hunk 1 above)
Comment (netoferraz) [reply airbyte-integrations/connectors/source-todoist/unit_tests/test_streams.py]: done.

**Comment by netoferraz** [comment]
Thank you @delenamalan for your review! I hope that I managed to fix and resolve all of your suggestions.

**Review by deanmalan — APPROVED** [review]
🎉

**Comment by delenamalan** [comment]
@sajarin could you advise on the next steps here, please? 🙏

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Comment by netoferraz** [comment]
> Hello wave:skin-tone-2: and thank you for your contribution!
> 
> Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. Because of this, reviewing and merging your contribution may take longer than usual. We apologize for the delay, but we want everyone to have a quiet and happy holiday.
> 
> If you have any questions or need further clarification, please don't hesitate to ping via Slack.

Hey @marcosmarxm Could we advance in this PR? Let me know if I can provide any further help here.

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**